### PR TITLE
[ShapesSpec] Support ShapesSpec.assumptions for cross-input shape invariants

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #185327
* #185165
* __->__ #185161
* #185154
* #184853

## What

Adds `ShapesSpec(assumptions=[...])` support — a list of `SymBool`
expressions built from spec `IntVar` / `ShapeVar` values that get
enforced at runtime AND fed to the ShapeEnv as compile-time facts, so
conditionals like `if x.shape[0] > y.shape[0]` resolve without DDE.

```python
A = ShapeVar("a")
B = ShapeVar("b")

def fn(x, y):
    if x.shape[0] > y.shape[0]:   # no DDE — assumption tells compiler the answer
        return x.sum() + 1000.0
    return y.sum() - 1000.0

torch.compile(
    fn,
    shapes_spec=ShapesSpec(
        params={"x": TensorSpec([A, None]), "y": TensorSpec([B, None])},
        assumptions=[A > B],
    ),
)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/fx/experimental/dynamic_spec.py [hunk 1]
```diff
@@ -236,7 +249,12 @@ class TensorSpec:
 
     def __init__(self, dims: Sequence[LeafIntSpec]) -> None:
         for i, d in enumerate(dims):
-            _validate_symint_spec(d, where=f"TensorSpec dim {i}")
+            if d is not None and not isinstance(d, (int, SymInt)):
+                raise TypeError(
+                    f"TensorSpec dim {i}: expected LeafIntSpec "
```
Comment (aorenste) [inline torch/fx/experimental/dynamic_spec.py]: (Claude Review) [cleanup] Error message is missing a connective: `f"TensorSpec dim {i}: expected LeafIntSpec {type(d).__name__}: {d!r}"` reads as `expected LeafIntSpec str: 'foo'`. Should be `expected LeafIntSpec, got {type(d).__name__}: {d!r}`.

File: torch/fx/experimental/dynamic_spec.py [hunk 2]
```diff
@@ -556,17 +573,36 @@ def __init__(
...
+                if not isinstance(a, SymBool):
+                    raise TypeError(
+                        f"ShapesSpec.assumptions[{i}]: expected SymBool, "
+                        f"got {type(a).__name__}: {a!r}"
+                    )
+                _validate_spec_sym(a, where=f"ShapesSpec.assumptions[{i}]")
+                self._assumptions.append(a)
+
+    @property
+    def assumptions(self) -> list[Any]:
```
Comment (aorenste) [inline torch/fx/experimental/dynamic_spec.py]: (Claude Review) [cleanup] Return annotation should be `list[SymBool]`, not `list[Any]` — the constructor validates each element is a `SymBool` (line 578) and `_assumptions` is typed as `list[SymBool]` (line 560), so the property is strictly typed.

File: test/dynamo/test_dynamic_spec.py [hunk 3]
```diff
@@ -973,6 +1031,152 @@ def fn(x, y, z):
...
+            ),
+        )
+        # Correct: A=5 > B=3, takes if branch.
+        out = compiled(torch.ones(5, 2), torch.ones(3, 2))
+        self.assertTrue(out.item() > 500.0)
+
+        # Violation: A=2 not > B=3.
+        torch._dynamo.reset()
+        with self.assertRaisesRegex(
+            (RuntimeError, AssertionError), "(Runtime assertion failed|Guard fail)"
```
Comment (aorenste) [inline test/dynamo/test_dynamic_spec.py]: (Claude Review) [question] Why the disjunction? Every other negative-path test in this file uses `assertRaisesRegex(AssertionError, "Guard fail")` (see lines 784, 834, 906, 982). A two-exception, two-message regex here suggests one of:

1. The implementation can legitimately fail two different ways depending on guard ordering — in which case the test passes by accident under one path and we lose coverage of the other.
2. The author was unsure which mode fires and over-specified to make the test pass — in which case the test will silently keep passing if the implementation regresses from "runtime assert" to "guard fail" (or vice versa) on this exact case.

Worth confirming which mode is the intended canonical failure and tightening the assertion to match. The substituted expression here is `Gt(u0, u1)` on unbacked symints, which I'd expect to go through `guard_or_defer_runtime_assert` and surface as the runtime-assert path — but the disjunction implies otherwise.

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -4381,20 +4392,37 @@ def _finalize_spec_wiring(shape_env: ShapeEnv) -> None:
...
+    for free, bool_expr in pending:
+        missing = free - subst_keys
+        if not missing:
+            raise RuntimeError(
+                f"_finalize_spec_wiring: pending entry has all symbols bound "
+                f"({bool_expr}); _drain_shape_spec_pending_assumptions should "
+                f"have removed it before finalize."
+            )
+        all_unbound |= missing
+        pretty_expr = re.sub(r"#\d+", "", str(bool_expr))
```
Comment (aorenste) [inline torch/_dynamo/variables/builder.py]: This seems like a superset of _pretty() - maybe move it there and call it here too?

File: torch/fx/experimental/dynamic_spec.py [hunk 5]
```diff
@@ -441,7 +447,7 @@ def func(x, n, *args, **kwargs): ...
 
     def __init__(
         self,
-        params: dict[str, IntermediateSpec] | None = None,
+        params: dict[str, Any] | None = None,
```
Comment (aorenste) [inline torch/fx/experimental/dynamic_spec.py]: Can this be a more specific type than 'Any'? If not maybe use `object` so you don't accidentally use it as unintended.

**Review by aorenste — APPROVED** [review]
Approving w/ comments. Claude's recommendation for more tests is probably apt.

File: torch/fx/experimental/dynamic_spec.py (hunk 2 above)
Comment (laithsakka) [reply torch/fx/experimental/dynamic_spec.py]: yep agree

File: torch/fx/experimental/dynamic_spec.py (hunk 5 above)
Comment (laithsakka) [reply torch/fx/experimental/dynamic_spec.py]: the type is  bit complicated thats why i made it any it could be List[IntermediateSpec] or IntermediateSpec or dict[str, IntermediateSpec]
maybe wroth making it explicit?

File: test/dynamo/test_dynamic_spec.py (hunk 3 above)
Comment (laithsakka) [reply test/dynamo/test_dynamic_spec.py]: ye this one need Runtime assertion failed because no guard is installed on unbacked well, all of them eventually shall be Runtime assertion failed but now we have a guard installed sometimes on unbacked (a known bug)

Comment (laithsakka) [reply test/dynamo/test_dynamic_spec.py]: i tightened it.

File: torch/_dynamo/variables/builder.py (hunk 4 above)
Comment (laithsakka) [reply torch/_dynamo/variables/builder.py]: addressed

File: torch/fx/experimental/dynamic_spec.py (hunk 5 above)
Comment (laithsakka) [reply torch/fx/experimental/dynamic_spec.py]: typed it

**Review by jansel — APPROVED** [review]
